docs/RFCS/20221124_split_tls.md

```diff
@@ -0,0 +1,295 @@
+own CA) for the HTTP service. However, more recently we have learned
+that the same concern exists also for the SQL service and the RPC
+service as used by CLI clients (cockroach node decommission, etc).
+
+Moving away from using the certificates used for node-to-node
+communication allows database operators to use their existing PKI (or
+automation such as LetsEncrypt) to manage the certificates their
+clients connect to, and simplifies future work to improve how the
+certificates used for node-to-node communication can be generated.
+
+# Technical design
+
+Currently the cert manager code inside CockroachDB has just two APIs
+for initializing+reading a 'server cert':
+
+- `GetUITLSconfig`, used by the HTTP service.
+- GetTLSConfig, used by all of:
+  - node-node RPC service
+  - node-client RPC service and RPC client
+  - node-client SQL service
+
+We will change this as follows:
+- `GetUITLSconfig` - unchanged
+- `GetTLSConfig`  gets split into:
+  - `GetRPCServerTLSConfig`, used for both node-node and node-client RPC service
+     Bundles 2 server peer certs in the TLSConfig, as specified below.
+  - `GetSQLServerTLSConfig`, used for the SQL service.
+  - `GetRPCClientTLSConfig`, used for RPC clients (client side of node-node connections, and also CLI clients)
+
+![server certs](20221124_split_tls/server_certs.png)
+
+If the `ui.crt` doesn’t exist, the HTTPS endpoint will fallback to the
+new `server.crt`, allowing database operators who don’t need a
+separation of those interfaces to use the same certificate for all
+three.
+
+The CA story is complementary to the story above: the `tls.Config`
+needs to know which CA cert to use to validate the _peer_ (the other
+side of the connection / the client). For each of the 3 services, and
+the RPC client, we bundle CAs as follows:
+
+![CA certs](20221124_split_tls/ca_certs.png)
+
+The diagrams above represent the desired end state of the
+functionality. Note that the code used to generate the `tls.Config` for
+SQL clients has been omitted: it is owned by the SQL client driver
+(e.g. pgx, lib/pq) and we don't control it.
+
+## Changes from current implementation
+
+To enumerate the actual changes, we also need to look at the current implementation.
+
+![server certs - orig](20221124_split_tls/server_certs_orig.png)
+
+![CA certs - orig](20221124_split_tls/ca_certs_orig.png)
+
+
+Enumeration of changes:
+
+- `GetTLSConfig` becomes split into:
+  - `GetSQLServerTLSConfig`
+  - `GetRPCServerTLSConfig`
+  - `GetRPCClientTLSConfig`
+
+- Each defined using the lookup logic defined above.
+
+- In particular:
+  - `ca-client.crt` becomes mandatory. There's a one-time copy of
+    ca.crt to ca-client.crt if ca-client.crt does not exist already.
+  - `server.crt` becomes mandatory. We also have a one-time copy from
+    node.crt.
+  - `ca-server.crt` is optional. It's possible to use the system trust
+    store to validate server.crt.
+  - `ca.crt` becomes optional: it becomes possible to use the system
```

Comment:
From a recommendation perspective, this is also a major change for what we currently ask customers. It makes sense to use what's in trust store for verified public CAs, but changing the documentation guideline may be confusing for certain existing customers, especially until we find time on docs team plate to make changes to existing docs. Is there a harm in still mandating the need to store ca.crt and the new ca-server.crt on cluster nodes?

Reply:
I can see the argument for mandating that `ca.crt` exist, but I'd argue requiring that `ca-server.crt` exist will cause problems for organizations that use LetsEncrypt, cloud provider managed certificates, etc, since they will have to track down the certificate tree that is signing their certificates and manually deploy them in `ca-server.crt`.

Reply:
Yeah agree.